only use mipmaps for power-of-two textures

ansis opened the pull request:
Fixes https://github.com/mapbox/mapbox-gl-js/issues/6440

The WMS source in that example returns 400x400 tiles. WebGL only supports mipmaps for power-of-two size textures. This fix just disables mipmap generation and filtering for textures that are not power-of-two.

Alternative: we could either upscale the image to a power-of-two texture or add padding and draw only part of the texture but both of these solutions are significantly more complex and might not look better. Mipmapping only helps rasters look a bit better when they are significantly underzoomed (zooming out and the tiles get smaller).

I wrote a render test for this but it doesn't fail without the fix (I guess the node module implementation doesn't error the same way) so I didn't include it here. I don't have any other ideas on how to test this.

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->

 - [x] briefly describe the changes in this PR
 - [tried] write tests for all new functionality
 - [n/a] document any changes to public APIs
 - [ ] post benchmark scores
 - [x] manually test the debug page

jfirebaugh commented on src/render/texture.js:
```diff
@@ -40,12 +40,14 @@ class Texture {
 
         const {width, height} = image;
         this.size = [width, height];
+        this.powerOfTwoSize = width === height && (Math.log(width) / Math.LN2) % 1 === 0;
```
Is there a reason to use state rather than compute this as needed? As state it could be wrong if the size is changed via `update`.

mourner replied:
BTW a common neat and fast way to check if a number is a power of two is `x & (x - 1) === 0`

ansis commented:
@jfirebaugh I moved setting `powerOfTwoSize` from the constructor to the `update` to fix the bug (
3936aaa). I can switch it to a method if you think that could avoid future bugs.

I also added 8b5ebb0 which moves update-without-reallocation logic into `Texture` so that we can handle that and mipmap generation within `Texture` instead of within `RasterTileSource`. I'm not completely sure if this is better of it's just less direct.

jfirebaugh commented on src/render/texture.js:
```diff
@@ -40,35 +40,51 @@ class Texture {
+    powerOfTwoSize: boolean;
 
-    constructor(context: Context, image: TextureImage, format: TextureFormat, premultiply: ?boolean) {
+    constructor(context: Context, image: TextureImage, format: TextureFormat, premultiply: ?boolean, useMipmap: ?boolean) {
```
Rather than add a second optional boolean parameter, let's move `premultiply` and `useMipmap` into an options object so they get descriptive names at the call site and there's no need for an `undefined` placeholder argument.
